### PR TITLE
[IOTDB-588] Add tags and attributes management

### PR DESCRIPTION
**Pull request by qiaojialin**
In this PR, we support:

create timeseries root.turbine.d1.s1(temperature) with datatype=FLOAT, encoding=RLE, compression=SNAPPY tags(unit=km, alarm=yes) attributes(group=g1)

show timeseries root.turbine.d1.s1 where unit=km

select temperature from root.turbine.d1

### CODE REVIEW

File: docs/zh/UserGuide/5-Operation Manual/1-DDL Data Definition Language.md [hunk 1]
```diff
@@ -71,6 +71,21 @@ error: encoding TS_2DIFF does not support BOOLEAN
...
+括号里的`temprature`是`s1`这个传感器的别名。
+我们可以在任何用到`s1`的地方，将其用`temprature`代替，这两者是等价的。
+
+> 注意：额外的标签和属性信息总的大小不能超过`tag_attribute_total_size`.
+
+标签和属性的唯一差别在于，我们为标签信息在内存中维护了一个倒排索引，所以可以在`show timeseries`的条件语句中使用标签作为查询条件，你将会在下一节看到具体查询内容。
+
 ## 查看时间序列
 
 目前，IoTDB支持两种查看时间序列的方式：
```
Comment (samperson1997) [inline docs/zh/UserGuide/5-Operation Manual/1-DDL Data Definition Language.md:91]: 三种

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 2]
```diff
@@ -454,6 +454,11 @@ partition_interval=604800
 memtable_num_in_each_storage_group=10
 
 
+# max size for tag and attribute of one time series
+# the unit is byte
+tag_attribute_total_size=700
```
Comment (samperson1997) [inline server/src/assembly/resources/conf/iotdb-engine.properties:459]: Should  we mention this config in document [chapter 3.4](https://iotdb.apache.org/zh/UserGuide/Master/3-Server/4-Config%20Manual.html)?

File: server/src/main/java/org/apache/iotdb/db/exception/metadata/AliasAlreadyExistException.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.apache.iotdb.rpc.TSStatusCode;
+
+public class AliasAlreadyExistException extends MetadataException {
+
+  private static final long serialVersionUID = 7299770003548114589L;
+
+  public AliasAlreadyExistException(String path, String alias) {
+    super(String.format("Alias [%s] for Path [%s] already exist", alias, path),
+            TSStatusCode.PATH_ALREADY_EXIST_ERROR.getStatusCode());
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/exception/metadata/AliasAlreadyExistException.java:31]: use the same error code as `PATH_ALREADY_EXIST_ERROR`?

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 4]
```diff
@@ -310,58 +337,36 @@ public void createTimeseries(String path, TSDataType dataType, TSEncoding encodi
...
-    try {
-      TSDataType tsDataType = TSDataType.valueOf(dataType);
-      TSEncoding tsEncoding = TSEncoding.valueOf(encoding);
-      CompressionType type = TSFileDescriptor.getInstance().getConfig().getCompressor();
-      createTimeseriesWithMemoryCheckAndLog(path, tsDataType, tsEncoding, type,
-          Collections.emptyMap());
-    } finally {
-      lock.writeLock().unlock();
-    }
+  public Set<LeafMNode> queryTimeseriesByTag(String tagKey, String tagValue) {
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:347]: This method is not used

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 5]
```diff
@@ -388,32 +357,74 @@ private QueryDataSet processShowStorageGroup() {
     return listDataSet;
   }
 
+  private QueryDataSet processShowTimeseriesWithIndex(ShowTimeSeriesPlan timeSeriesPlan) throws MetadataException {
+    List<ShowTimeSeriesResult> timeseriesList = MManager.getInstance()
+            .getAllMeasurementSchema(timeSeriesPlan.getPath().toString(),
+                    timeSeriesPlan.isContains(), timeSeriesPlan.getKey(), timeSeriesPlan.getValue());
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:363]: Should we just pass `timeSeriesPlan` as param of getAllMeasurementSchema()  method? As `createTimeseries(createTimeSeriesPlan)` did.

File: server/src/main/java/org/apache/iotdb/db/exception/metadata/AliasAlreadyExistException.java (hunk 3 above)
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/exception/metadata/AliasAlreadyExistException.java:31]: added a ALIAS_ALREADY_EXIST_ERROR

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java (hunk 4 above)
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:347]: removed

File: docs/zh/UserGuide/5-Operation Manual/1-DDL Data Definition Language.md (hunk 1 above)
Comment (qiaojialin) [reply docs/zh/UserGuide/5-Operation Manual/1-DDL Data Definition Language.md:91]: fixed

File: server/src/assembly/resources/conf/iotdb-engine.properties (hunk 2 above)
Comment (qiaojialin) [reply server/src/assembly/resources/conf/iotdb-engine.properties:459]: added

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java (hunk 5 above)
Comment (qiaojialin) [reply server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:363]: fixed

**Comment by qiaojialin** [comment]
> Good job! Just a few comments : )
> By the way, when the size of the extra tag and attribute information exceed `tag_attribute_total_size`, is there just an exception thrown? It may hint the users to enlarge the value of `tag_attribute_total_size`, but I think it is a little bit unkind?🤔

added the hint

File: server/src/test/java/org/apache/iotdb/db/engine/modification/DeletionFileNodeTest.java [hunk 6]
```diff
@@ -27,6 +27,7 @@
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import org.apache.iotdb.db.conf.directories.DirectoryManager;
```
Comment (HTHou) [inline server/src/test/java/org/apache/iotdb/db/engine/modification/DeletionFileNodeTest.java:32]: Should we put these java imports under the iotdb packages?

File: pom.xml [hunk 7]
```diff
@@ -115,7 +115,7 @@
         <thrift.version>0.13.0</thrift.version>
         <airline.version>0.8</airline.version>
         <jackson.version>2.10.0</jackson.version>
-        <antlr4.version>4.7.1</antlr4.version>
+        <antlr4.version>4.8-1</antlr4.version>
         <common.cli.version>1.3.1</common.cli.version>
```
[the remark this answers is not in the document]
Comment (qiaojialin) [reply pom.xml:119]: fixed

**Pull request merged**